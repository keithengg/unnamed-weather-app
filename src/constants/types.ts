export type Weather = {
    temp: number;
    temp_max: number;
    temp_min: number;
    humidity: number;
    city: string;
    time: string;
    cloud_status?: string;
}