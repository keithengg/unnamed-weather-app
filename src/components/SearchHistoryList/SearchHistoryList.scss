.search-history {
    &__item {
        display: flex;
        justify-content: space-between;
        background-color: rgba(355, 355, 355, 0.3);
        border-radius: 15px;
        border: 1px solid #f2f2f2;
        padding: 16px 8px 16px;
        margin-top: 16px;
    }

    &__button {
        background-color: #f2f2f2 !important;
        padding: 16px;
        border-radius: 100%;
    }
}