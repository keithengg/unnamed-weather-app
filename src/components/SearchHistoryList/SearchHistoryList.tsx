import React from "react";
import { History } from '../../constants/types';

import { HiOutlineMagnifyingGlass } from "react-icons/hi2";
import { FaTrash } from "react-icons/fa";

import './SearchHistoryList.scss';

interface SearchHistoryListProps {
    searchHistory: History[];
    handleDelete: (index: number) => void;
    handleSearch: (searchLocation: string) => void;
};

interface SearchHistoryProps {
    history: History;
    searchHistory: () => void;
    deleteHistory: () => void;
};

const SearchHistory: React.FC<SearchHistoryProps> = ({ history, searchHistory, deleteHistory }) => {
    return (
        <li className="search-history__item">
            <div className="flex flex-col">
                <p>{history.city}</p>
                <p>{history.time}</p>
            </div>
            <div className="flex gap-x-2">
                <button
                    className="search-history__button"
                    onClick={searchHistory}
                    type="button"
                    aria-label={`Search for ${history.city}`}
                >
                    <HiOutlineMagnifyingGlass size={24} />
                </button>
                <button
                    className="search-history__button"
                    onClick={deleteHistory}
                    type="button"
                    aria-label={`Delete ${history.city} history`}
                >
                    <FaTrash size={24} />
                </button>
            </div>
        </li>
    )
}

const SearchHistoryList: React.FC<SearchHistoryListProps> = ({ searchHistory, handleDelete, handleSearch }) => {
    return (
        <div>
            Search History
            <ul className="search-history__list">
                {searchHistory && searchHistory.map((history, index) => {
                    return (
                        <SearchHistory
                            key={`${history.city}${history.time}`}
                            history={history}
                            searchHistory={() => handleSearch(history.city)}
                            deleteHistory={() => handleDelete(index)}
                        />
                    )
                })}
            </ul>
        </div>
    )
}

export default SearchHistoryList; // Note to marker: Thought about using React.Memo here and useCallback on the "handleDelete/Search" functions but it seems like overkill