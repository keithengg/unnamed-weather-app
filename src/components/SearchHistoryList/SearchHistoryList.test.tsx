import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import '@testing-library/jest-dom'

import SearchHistoryList from "./SearchHistoryList";
import { History } from '../../constants/types'

describe("<SearchHistoryList />", () => {
    const mockHandleDelete = jest.fn();
    const mockHandleSearch = jest.fn();

    const mockData: History[] = [
        {
            city: "Singapore",
            time: "9:00AM"
        },
        {
            city: "Malaysia",
            time: "10:00AM"
        },
        {
            city: "Thailand",
            time: "11:00AM"
        },
        {
            city: "Vietnam",
            time: "12:00PM"
        }
    ]

    it('should render correctly and match snapshot', () => {
        const { container } = render(
            <SearchHistoryList
                searchHistory={mockData}
                handleSearch={mockHandleSearch}
                handleDelete={mockHandleDelete}
            />
        )
        expect(container).toMatchSnapshot()
    });

    it('should render and have all relevant search history data', () => {
        render(
            <SearchHistoryList
                searchHistory={mockData}
                handleSearch={mockHandleSearch}
                handleDelete={mockHandleDelete}
            />
        )

        for (let i = 0; i < mockData.length; i++) {
            expect(screen.getByText(mockData[i].city)).toBeInTheDocument();
            expect(screen.getByText(mockData[i].time)).toBeInTheDocument();

        }
    });

    it('should call appriopriate search function with correct arguments', () => {
        render(
            <SearchHistoryList
                searchHistory={mockData}
                handleSearch={mockHandleSearch}
                handleDelete={mockHandleDelete}
            />
        )

        const searchButton = screen.getByRole('button', { name: `Search for ${mockData[0].city}` });
        expect(searchButton).toBeInTheDocument();

        fireEvent.click(searchButton);
        expect(mockHandleSearch).toHaveBeenCalledWith(mockData[0].city)
        expect(mockHandleSearch).toHaveBeenCalledTimes(1)
    });


    it('should call appriopriate delete function with correct arguments', () => {
        render(
            <SearchHistoryList
                searchHistory={mockData}
                handleSearch={mockHandleSearch}
                handleDelete={mockHandleDelete}
            />
        )

        const deleteButton = screen.getByRole('button', { name: `Delete ${mockData[3].city} history` });
        expect(deleteButton).toBeInTheDocument();

        fireEvent.click(deleteButton);
        expect(mockHandleDelete).toHaveBeenCalledWith(3)
        expect(mockHandleDelete).toHaveBeenCalledTimes(1)
    });
});