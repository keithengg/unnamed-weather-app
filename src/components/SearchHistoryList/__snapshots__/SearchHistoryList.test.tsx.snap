// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchHistoryList /> should render correctly and match snapshot 1`] = `
<div>
  <div>
    Search History
    <ul>
      <li
        class="search-history__item"
      >
        <div
          class="flex flex-col"
        >
          <p>
            Singapore
          </p>
          <p>
            9:00AM
          </p>
        </div>
        <div
          class="flex gap-x-2"
        >
          <button
            aria-label="Search for Singapore"
            class="search-history__button"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21 21l-5.197-5.197m0 0A7.5 7.5 0 105.196 5.196a7.5 7.5 0 0010.607 10.607z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
          <button
            aria-label="Delete Singapore history"
            class="search-history__button"
            type="button"
          >
            <svg
              fill="currentColor"
              height="24"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
              />
            </svg>
          </button>
        </div>
      </li>
      <li
        class="search-history__item"
      >
        <div
          class="flex flex-col"
        >
          <p>
            Malaysia
          </p>
          <p>
            10:00AM
          </p>
        </div>
        <div
          class="flex gap-x-2"
        >
          <button
            aria-label="Search for Malaysia"
            class="search-history__button"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21 21l-5.197-5.197m0 0A7.5 7.5 0 105.196 5.196a7.5 7.5 0 0010.607 10.607z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
          <button
            aria-label="Delete Malaysia history"
            class="search-history__button"
            type="button"
          >
            <svg
              fill="currentColor"
              height="24"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
              />
            </svg>
          </button>
        </div>
      </li>
      <li
        class="search-history__item"
      >
        <div
          class="flex flex-col"
        >
          <p>
            Thailand
          </p>
          <p>
            11:00AM
          </p>
        </div>
        <div
          class="flex gap-x-2"
        >
          <button
            aria-label="Search for Thailand"
            class="search-history__button"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21 21l-5.197-5.197m0 0A7.5 7.5 0 105.196 5.196a7.5 7.5 0 0010.607 10.607z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
          <button
            aria-label="Delete Thailand history"
            class="search-history__button"
            type="button"
          >
            <svg
              fill="currentColor"
              height="24"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
              />
            </svg>
          </button>
        </div>
      </li>
      <li
        class="search-history__item"
      >
        <div
          class="flex flex-col"
        >
          <p>
            Vietnam
          </p>
          <p>
            12:00PM
          </p>
        </div>
        <div
          class="flex gap-x-2"
        >
          <button
            aria-label="Search for Vietnam"
            class="search-history__button"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21 21l-5.197-5.197m0 0A7.5 7.5 0 105.196 5.196a7.5 7.5 0 0010.607 10.607z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
          <button
            aria-label="Delete Vietnam history"
            class="search-history__button"
            type="button"
          >
            <svg
              fill="currentColor"
              height="24"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
              />
            </svg>
          </button>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
